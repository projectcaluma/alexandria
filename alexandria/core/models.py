import re
import uuid

from django.core.exceptions import ImproperlyConfigured
from django.core.validators import RegexValidator
from django.db import models
from django.utils.translation import gettext_lazy as _
from localized_fields.fields import LocalizedCharField, LocalizedTextField
from rest_framework import exceptions

from .storage_clients import client


def make_uuid():
    """Return a new random UUID value.

    This indirection is done for testing purposes, so test code can mock
    uuid.uuid4(). If we wouldn't do this, then the models would have a direct
    reference that doesn't get mocked away.

    We can't replace it with a lambda because Django Migrations can't handle them.
    """
    return uuid.uuid4()


class VisibilityMixin:
    visibility_classes = None

    @classmethod
    def visibility_queryset_filter(cls, queryset, request, **kwargs):
        if cls.visibility_classes is None:
            raise ImproperlyConfigured(
                "check that app `alexandria.core.apps.DefaultConfig` is part of your `INSTALLED_APPS`."
            )

        for visibility_class in cls.visibility_classes:
            queryset = visibility_class().filter_queryset(cls, queryset, request)

        return queryset


class PermissionMixin:
    permission_classes = None

    @classmethod
    def check_permissions(cls, request, **kwargs):
        if cls.permission_classes is None:
            raise ImproperlyConfigured(
                "check that app `alexandria.core.apps.DefaultConfig` is part of your `INSTALLED_APPS`."
            )

        for permission_class in cls.permission_classes:
            if not permission_class().has_permission(cls, request):
                raise exceptions.PermissionDenied()

    def check_object_permissions(self, request):
        for permission_class in self.permission_classes:
            if not permission_class().has_object_permission(
                self.__class__, request, self
            ):
                raise exceptions.PermissionDenied()


class BaseModel(PermissionMixin, VisibilityMixin, models.Model):
    created_at = models.DateTimeField(auto_now_add=True, db_index=True)
    created_by_user = models.CharField(
        _("created by user"), max_length=150, blank=True, null=True
    )
    created_by_group = models.CharField(
        _("created by group"), max_length=255, blank=True, null=True
    )
    modified_at = models.DateTimeField(auto_now=True, db_index=True)
    modified_by_user = models.CharField(
        _("created by user"), max_length=150, blank=True, null=True
    )
    modified_by_group = models.CharField(
        _("created by group"), max_length=255, blank=True, null=True
    )
    metainfo = models.JSONField(_("metainfo"), default=dict)

    class Meta:
        abstract = True


class UUIDModel(BaseModel):
    """
    Models which use uuid as primary key.

    Defined as alexandria default
    """

    id = models.UUIDField(primary_key=True, default=make_uuid, editable=False)

    class Meta:
        abstract = True


class SlugModel(BaseModel):
    """
    Models which use a slug as primary key.

    Defined as alexandria default for configuration so it is possible
    to merge between developer and user configuration.
    """

    slug = models.SlugField(max_length=255, primary_key=True)

    class Meta:
        abstract = True


COLOR_RE = re.compile("^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$")
color_validator = RegexValidator(COLOR_RE, _("Enter a valid color."), "invalid")


class Category(SlugModel):
    name = LocalizedCharField(
        _("category name"), blank=False, null=False, required=False
    )
    description = LocalizedTextField(
        _("category description"), null=True, blank=True, required=False
    )
    color = models.CharField(
        max_length=18,
        default="#FFFFFF",
        validators=[color_validator],
    )
    parent = models.ForeignKey(
        "self",
        on_delete=models.CASCADE,
        blank=True,
        null=True,
        related_name="children",
    )


class TagSynonymGroup(BaseModel):
    pass


class Tag(SlugModel):
    name = models.CharField(_("tag name"), blank=False, null=False, max_length=100)
    description = LocalizedTextField(
        _("tag description"), null=True, blank=True, required=False
    )
    tag_synonym_group = models.ForeignKey(
        TagSynonymGroup,
        on_delete=models.SET_NULL,
        blank=True,
        null=True,
        related_name="tags",
    )


class Mark(SlugModel):
    name = LocalizedTextField(_("Mark name"), blank=False, null=False, max_length=100)
    description = LocalizedTextField(
        _("Mark description"), null=True, blank=True, required=False
    )


class Document(UUIDModel):
    title = LocalizedCharField(
        _("document title"), blank=True, null=True, required=False
    )
    description = LocalizedTextField(
        _("document description"), null=True, blank=True, required=False
    )
    category = models.ForeignKey(
        Category,
        on_delete=models.SET_NULL,
        null=True,
        blank=True,
        related_name="documents",
    )
    tags = models.ManyToManyField(Tag, blank=True, related_name="documents")
    marks = models.ManyToManyField(Mark, blank=True, related_name="documents")
    date = models.DateField(blank=True, null=True)

    def clone(self):
        latest_original = (
            self.files.filter(variant="original").order_by("-created_at").first()
        )
        self.pk = None
        self.save()

        # no need to clone the thumbnail, it will be auto-created
        new_original = latest_original.clone()
        new_original.document = self
        new_original.save()

        return self


class File(UUIDModel):
    ORIGINAL = "original"
    THUMBNAIL = "thumbnail"

    VARIANT_CHOICES = (
        ORIGINAL,
        THUMBNAIL,
    )
    VARIANT_CHOICES_TUPLE = (
        (variant_choice, variant_choice) for variant_choice in VARIANT_CHOICES
    )
    variant = models.CharField(
        choices=VARIANT_CHOICES_TUPLE, max_length=23, default=ORIGINAL
    )
    original = models.ForeignKey(
        "self",
        on_delete=models.CASCADE,
        null=True,
        blank=True,
        related_name="renderings",
    )
    name = models.CharField(_("file name"), max_length=255)
    document = models.ForeignKey(
        Document, on_delete=models.CASCADE, related_name="files"
    )
    checksum = models.CharField(_("checksum"), max_length=255, null=True, blank=True)

    UNDEFINED = "undefined"
    COMPLETED = "completed"
    ERROR = "error"

    STATUS_CHOICES = (
        UNDEFINED,
        COMPLETED,
        ERROR,
    )
    UPLOAD_STATUS_TUPLE = tuple(
        [(status_choice, status_choice) for status_choice in STATUS_CHOICES]
    )
    upload_status = models.CharField(
        choices=UPLOAD_STATUS_TUPLE, max_length=32, default=UNDEFINED
    )

    class Meta:
        ordering = ["-created_at"]

    def delete(self, *args, **kwargs):
        super().delete(*args, **kwargs)
        client.remove_object(self.object_name)

    @property
    def object_name(self):
        return f"{self.pk}_{self.name}"

    @property
    def upload_url(self):
        return client.upload_url(self.object_name)

    @property
    def download_url(self):
        return client.download_url(self.object_name)

    def clone(self):
        source_name = self.object_name
        self.pk = None
        self.save()
        client.copy_object(source_name, self.object_name)
        return self
