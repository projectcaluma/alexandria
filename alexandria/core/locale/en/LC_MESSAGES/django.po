# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-02-21 09:05+0000\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"

#: alexandria/core/api.py
msgid "(copy)"
msgstr ""

#: alexandria/core/models.py
msgid "Created by user"
msgstr ""

#: alexandria/core/models.py
msgid "Created by group"
msgstr ""

#: alexandria/core/models.py
msgid "Modified by user"
msgstr ""

#: alexandria/core/models.py
msgid "Modified by group"
msgstr ""

#: alexandria/core/models.py
msgid "Metainfo"
msgstr ""

#: alexandria/core/models.py
msgid "Please enter a valid HEX color code."
msgstr ""

#: alexandria/core/models.py
msgid "Name"
msgstr ""

#: alexandria/core/models.py
msgid "Description"
msgstr ""

#: alexandria/core/models.py
msgid "Allowed mime types"
msgstr ""

#: alexandria/core/models.py
msgid "Title"
msgstr ""

#: alexandria/core/models.py
msgid "Checksum"
msgstr ""

#: alexandria/core/validations.py
msgid "File is infected with malware."
msgstr ""

#: alexandria/core/validations.py
msgid "Malware scan had an error: "
msgstr ""

#: alexandria/core/validations.py
#, python-format
msgid "File type %(mime_type)s is not allowed in category %(category)s."
msgstr ""

#: alexandria/core/validations.py
msgid "Missing Content-Type header"
msgstr ""

#: alexandria/core/validations.py
msgid "Unknown file extension"
msgstr ""

#: alexandria/core/validations.py
#, python-format
msgid ""
"Content-Type %(content_type)s does not match file extension %(extension)s."
msgstr ""

#: alexandria/core/validations.py
#, python-format
msgid ""
"Content-Type %(content_type)s does not match detected file content "
"%(file_content_type)s."
msgstr ""

#: alexandria/core/views.py
msgid "PDF conversion is not enabled."
msgstr ""

#: alexandria/core/views.py
msgid "\"files\" filter is mandatory!"
msgstr ""

#: alexandria/core/views.py
msgid "The \"files\" filter must consist of a comma delimited list of UUIDs!"
msgstr ""

#: alexandria/core/views.py
msgid "For downloading a file use the presigned download URL."
msgstr ""

#: alexandria/core/views.py
msgid "WebDAV is not enabled."
msgstr ""

#: alexandria/core/views.py
msgid "WebDAV is not enabled for this mime type."
msgstr ""
