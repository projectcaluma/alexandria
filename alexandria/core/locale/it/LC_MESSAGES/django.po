# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-02-21 09:05+0000\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"

#: alexandria/core/api.py
msgid "(copy)"
msgstr "(copia)"

#: alexandria/core/models.py
msgid "Created by user"
msgstr "Creato da utente"

#: alexandria/core/models.py
msgid "Created by group"
msgstr "Creato da gruppo"

#: alexandria/core/models.py
msgid "Modified by user"
msgstr "Modificato da utente"

#: alexandria/core/models.py
msgid "Modified by group"
msgstr "Modificato  da gruppo"

#: alexandria/core/models.py
msgid "Metainfo"
msgstr "Metadati"

#: alexandria/core/models.py
msgid "Please enter a valid HEX color code."
msgstr "Inserire un codice colore HEX valido."

#: alexandria/core/models.py
msgid "Name"
msgstr "Nome"

#: alexandria/core/models.py
msgid "Description"
msgstr "Descrizione"

#: alexandria/core/models.py
msgid "Allowed mime types"
msgstr "Tipi di file consentiti"

#: alexandria/core/models.py
msgid "Title"
msgstr "Titolo"

#: alexandria/core/models.py
msgid "Checksum"
msgstr "Checksum"

#: alexandria/core/validations.py
msgid "File is infected with malware."
msgstr "Il file è infetto da malware."

#: alexandria/core/validations.py
msgid "Malware scan had an error: "
msgstr "La scansione del malware ha avuto un errore: "

#: alexandria/core/validations.py
#, python-format
msgid "File type %(mime_type)s is not allowed in category %(category)s."
msgstr ""
"Il tipo di file %(mime_type)s non è consentito nella categoria %(category)s."

#: alexandria/core/validations.py
msgid "Missing Content-Type header"
msgstr "Intestazione Content-Type mancante"

#: alexandria/core/validations.py
msgid "Unknown file extension"
msgstr "Estensione file sconosciuta"

#: alexandria/core/validations.py
#, python-format
msgid ""
"Content-Type %(content_type)s does not match file extension %(extension)s."
msgstr ""
"Il tipo di contenuto %(content_type)s non corrisponde all'estensione "
"%(extension)s del file."

#: alexandria/core/validations.py
#, python-format
msgid ""
"Content-Type %(content_type)s does not match detected file content "
"%(file_content_type)s."
msgstr ""
"Il tipo di contenuto %(content_type)s non corrisponde al contenuto del file "
"%(file_content_type)s."

#: alexandria/core/views.py
msgid "PDF conversion is not enabled."
msgstr "La conversione PDF non è attivata."

#: alexandria/core/views.py
msgid "\"files\" filter is mandatory!"
msgstr "Il filtro \"files\" è obbligatorio!"

#: alexandria/core/views.py
msgid "The \"files\" filter must consist of a comma delimited list of UUIDs!"
msgstr ""
"Il filtro \"files\" deve essere costituito da un elenco di UUID delimitato "
"da virgole!"

#: alexandria/core/views.py
msgid "For downloading a file use the presigned download URL."
msgstr "Per scaricare un file, utilizzare l'URL di download predefinito."

#: alexandria/core/views.py
msgid "WebDAV is not enabled."
msgstr "WebDAV non è attivato."

#: alexandria/core/views.py
msgid "WebDAV is not enabled for this mime type."
msgstr "WebDAV non è abilitato per questo tipo di file."
