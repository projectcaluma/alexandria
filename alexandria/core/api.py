import logging

from django.conf import settings
from django.core.exceptions import ValidationError
from django.core.files import File

from alexandria.core import models

log = logging.getLogger(__name__)


def create_document_file(
    user: str,
    group: str,
    category: models.Category,
    document_title: str,
    file_name: str,
    file_content: File,
    mime_type: str,
    file_size: int,
    additional_document_attributes={},
    additional_file_attributes={},
):
    """
    Create a document and file with the given attributes.

    This function eases the creation of documents and files by automatically setting important fields.
    Uses `create_file` to create the file.
    """
    document = models.Document.objects.create(
        title=document_title,
        category=category,
        created_by_user=user,
        created_by_group=group,
        modified_by_user=user,
        modified_by_group=group,
        **additional_document_attributes,
    )
    file = create_file(
        document=document,
        user=user,
        group=group,
        name=file_name,
        content=file_content,
        mime_type=mime_type,
        size=file_size,
        **additional_file_attributes,
    )

    return document, file


def create_file(
    document: models.Document,
    user: str,
    group: str,
    name: str,
    content: File,
    mime_type: str,
    size: int,
    **additional_attributes
):
    """
    Create a file with defaults and generate a thumbnail.

    Use this instead of the normal File.objects.create to ensure that all important fields are set.
    As well as generating a thumbnail for the file.
    """
    file = models.File.objects.create(
        name=name,
        content=content,
        mime_type=mime_type,
        size=size,
        document=document,
        encryption_status=(
            settings.ALEXANDRIA_ENCRYPTION_METHOD
            if settings.ALEXANDRIA_ENABLE_AT_REST_ENCRYPTION
            else None
        ),
        created_by_user=user,
        created_by_group=group,
        modified_by_user=user,
        modified_by_group=group,
        **additional_attributes,
    )

    try:
        file.create_thumbnail()
    except ValidationError as e:  # pragma: no cover
        log.error(
            "Object {obj} created successfully. Thumbnail creation failed. Error: {error}".format(
                obj=file, error=e.messages
            )
        )

    return file
