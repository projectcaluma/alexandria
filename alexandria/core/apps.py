from django.apps import AppConfig


class DefaultConfig(AppConfig):
    name = "alexandria.core"
    label = "alexandria_core"
