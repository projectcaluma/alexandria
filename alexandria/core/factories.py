from factory import Faker, SubFactory, post_generation
from factory.django import DjangoModelFactory

from . import models


class BaseFactory(DjangoModelFactory):
    created_by_user = "admin"
    created_by_group = "admin"
    modified_by_user = "admin"
    modified_by_group = "admin"

    class Meta:
        abstract = True


class CategoryFactory(BaseFactory):
    slug = Faker("slug")
    name = Faker("name")
    description = Faker("text")
    color = Faker("color")
    metainfo = {}

    class Meta:
        model = models.Category


class TagSynonymGroupFactory(BaseFactory):
    class Meta:
        model = models.TagSynonymGroup

    @post_generation
    def tags(self, create, extracted, **kwargs):
        if not create:  # pragma: todo cover
            # Simple build, do nothing.
            return

        if extracted:
            # A list of tags were passed in, use them
            for tag in extracted:
                self.tags.add(tag)


class TagFactory(BaseFactory):
    slug = Faker("slug")
    name = Faker("name")
    description = Faker("text")

    class Meta:
        model = models.Tag


class MarkFactory(BaseFactory):
    slug = Faker("slug")
    name = Faker("name")
    description = Faker("text")

    class Meta:
        model = models.Mark


class DocumentFactory(BaseFactory):
    class Meta:
        model = models.Document

    title = Faker("name")
    description = Faker("text")
    category = SubFactory(CategoryFactory)
    date = Faker("date")

    @post_generation
    def tags(self, create, extracted, **kwargs):
        if not create:  # pragma: todo cover
            # Simple build, do nothing.
            return

        if extracted:
            # A list of tags were passed in, use them
            for tag in extracted:
                self.tags.add(tag)

    @post_generation
    def marks(self, create, extracted, **kwargs):  # pragma: todo cover
        if not create:
            return

        if extracted:
            for mark in extracted:
                self.marks.add(mark)


class FileFactory(BaseFactory):
    name = Faker("name")
    document = SubFactory(DocumentFactory)

    class Meta:
        model = models.File


class DocumentTagsFactory(DjangoModelFactory):
    document = SubFactory(DocumentFactory)
    tag = SubFactory(TagFactory)

    class Meta:
        model = models.Document.tags.through


class DocumentMarksFactory(DjangoModelFactory):
    document = SubFactory(DocumentFactory)
    mark = SubFactory(MarkFactory)

    class Meta:
        model = models.Document.marks.through
