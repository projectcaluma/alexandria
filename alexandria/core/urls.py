from django.urls import path
from rest_framework.routers import SimpleRouter

from . import views

r = SimpleRouter(trailing_slash=False)

r.register(r"categories", views.CategoryViewSet)
r.register(r"documents", views.DocumentViewSet)
r.register(r"files", views.FileViewSet)
r.register(r"tags", views.TagViewSet)
r.register(r"marks", views.MarkViewSet)
r.register(r"tagsynonymgroups", views.TagSynonymGroupViewSet)

urlpatterns = r.urls

urlpatterns.append(path("hook", views.hook_view, name="hook"))
