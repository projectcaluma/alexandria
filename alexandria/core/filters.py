import json

from django.contrib.postgres.fields.jsonb import KeyTextTransform
from django.db.models import FloatField, Q, TextField, Value
from django.db.models.functions import Cast
from django_filters import (
    BaseCSVFilter,
    BaseInFilter,
    BooleanFilter,
    CharFilter,
    Filter,
    FilterSet,
)
from django_filters.constants import EMPTY_VALUES
from rest_framework.exceptions import ValidationError

from alexandria.core import models


class CharInFilter(BaseInFilter, CharFilter):
    pass


class JSONValueFilter(Filter):
    field_type_by_lookup_expr = {
        "exact": TextField,
        "gt": FloatField,
        "lt": FloatField,
        "gte": FloatField,
        "lte": FloatField,
    }

    def filter(self, qs, value):
        if value in EMPTY_VALUES:
            return qs

        valid_lookups = self._valid_lookups(qs)

        try:
            value = json.loads(value)
        except json.decoder.JSONDecodeError:
            raise ValidationError("JSONValueFilter value needs to be json encoded.")

        if isinstance(value, dict):
            # be a bit more tolerant
            value = [value]

        for expr in value:
            if expr in EMPTY_VALUES:  # pragma: no cover
                continue
            if not all(("key" in expr, "value" in expr)):
                raise ValidationError(
                    'JSONValueFilter value needs to have a "key" and "value" and an '
                    'optional "lookup" key.'
                )

            lookup_expr = expr.get("lookup", self.lookup_expr)
            if lookup_expr not in valid_lookups:
                raise ValidationError(
                    f'Lookup expression "{lookup_expr}" not allowed for field '
                    f'"{self.field_name}". Valid expressions: '
                    f'{", ".join(valid_lookups.keys())}'
                )
            # "contains" behaves differently on JSONFields as it does on other fields.
            # That's why we annotate the queryset with the value.
            # Some discussion about it can be found here:
            # https://code.djangoproject.com/ticket/26511
            cast_to = self.field_type_by_lookup_expr.get(lookup_expr, TextField)
            qs = qs.annotate(
                field_val=Cast(
                    KeyTextTransform(expr["key"], self.field_name),
                    output_field=cast_to(),
                )
            )
            lookup = {f"field_val__{lookup_expr}": Value(expr["value"])}

            qs = qs.filter(**lookup)
        return qs

    def _valid_lookups(self, qs):
        # We need some traversal magic in case field name is a related lookup
        traversals = self.field_name.split("__")
        actual_field = traversals.pop()

        model = qs.model
        for field in traversals:
            model = model._meta.get_field(field).related_model

        return model._meta.get_field(actual_field).get_lookups()


class ActiveGroupFilter(CharFilter):
    def filter(self, qs, value):
        if value in EMPTY_VALUES:
            return qs
        if value not in self.parent.request.user.groups:
            raise ValidationError(
                f"Active group '{value}' is not part of user's assigned groups"
            )
        return qs


class TagsFilter(BaseInFilter):
    def filter(self, qs, value):
        if value in EMPTY_VALUES:
            return qs
        # Documents must have all given tags
        # including each tag's synonyms
        for tag in value:
            synonyms = models.Tag.objects.filter(
                Q(pk=tag) | Q(tag_synonym_group__tags=tag)
            )
            qs = qs.filter(tags__in=synonyms)
        return qs


class CategoriesFilter(Filter):
    def filter(self, qs, value):
        if value in EMPTY_VALUES:
            return qs

        exclude_children = self.parent.request.query_params.get(
            "filter[exclude_children]", False
        )
        filters = Q()

        for category in value.split(","):
            filters |= Q(**{f"{self.field_name}_id": category})
            if not exclude_children:
                filters |= Q(**{f"{self.field_name}__parent__slug": category})

        return qs.filter(filters)


class CategoryFilterSet(FilterSet):
    metainfo = JSONValueFilter(field_name="metainfo")
    active_group = ActiveGroupFilter()
    has_parent = BooleanFilter(field_name="parent", lookup_expr="isnull", exclude=True)
    slug = CharFilter()
    slugs = CharInFilter(field_name="slug", lookup_expr="in")

    class Meta:
        model = models.Category
        fields = ["active_group", "metainfo", "has_parent", "slug", "slugs"]


class DocumentFilterSet(FilterSet):
    metainfo = JSONValueFilter(field_name="metainfo")
    active_group = ActiveGroupFilter()
    tags = TagsFilter()
    marks = CharInFilter()
    category = CategoriesFilter()
    categories = CategoriesFilter(field_name="category")
    # exclude_children is applied in CategoriesFilter, this is needed for DjangoFilterBackend
    exclude_children = BooleanFilter(field_name="title", method=lambda qs, __, ___: qs)

    class Meta:
        model = models.Document
        fields = ["metainfo", "tags"]


class FileFilterSet(FilterSet):
    metainfo = JSONValueFilter(field_name="metainfo")
    document_metainfo = JSONValueFilter(field_name="document__metainfo")
    active_group = ActiveGroupFilter()
    files = BaseCSVFilter(field_name="pk", lookup_expr="in")

    class Meta:
        model = models.File
        fields = ["original", "renderings", "variant", "metainfo", "files"]


class TagFilterSet(FilterSet):
    metainfo = JSONValueFilter(field_name="metainfo")
    active_group = ActiveGroupFilter()
    with_documents_in_category = CategoriesFilter(field_name="documents__category")
    with_documents_metainfo = JSONValueFilter(field_name="documents__metainfo")
    name = CharFilter(lookup_expr="istartswith")
    name_exact = CharFilter(field_name="name", lookup_expr="iexact")
    # exclude_children is applied in CategoriesFilter, this is needed for DjangoFilterBackend
    exclude_children = BooleanFilter(field_name="name", method=lambda qs, __, ___: qs)

    class Meta:
        model = models.Tag
        fields = [
            "metainfo",
            "active_group",
            "with_documents_in_category",
            "with_documents_metainfo",
            "name",
        ]


class MarkFilterSet(FilterSet):
    metainfo = JSONValueFilter(field_name="metainfo")
    active_group = ActiveGroupFilter()
    with_documents_in_category = CharFilter(field_name="documents__category__slug")
    with_documents_metainfo = JSONValueFilter(field_name="documents__metainfo")
    name = CharFilter(lookup_expr="istartswith")

    class Meta:
        model = models.Mark
        fields = [
            "metainfo",
            "active_group",
            "with_documents_in_category",
            "with_documents_metainfo",
            "name",
        ]
