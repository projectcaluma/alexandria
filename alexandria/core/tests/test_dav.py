from pathlib import Path
from uuid import uuid4
from xml.dom import minidom

import boto3
import pytest
from django.core.exceptions import ValidationError
from django.core.files.base import ContentFile
from django.urls import reverse
from manabi.token import Key, Token
from moto import mock_aws
from rest_framework import status
from rest_framework.status import HTTP_200_OK
from webtest import TestApp, TestRequest
from wsgidav.dav_error import HTTP_FORBIDDEN

from alexandria.core.models import File
from alexandria.dav import get_dav
from alexandria.dav_provider import AlexandriaProvider


@pytest.fixture
def s3(settings):
    with mock_aws():
        return boto3.client(
            "s3",
            endpoint_url=settings.ALEXANDRIA_S3_ENDPOINT_URL,
            aws_access_key_id=settings.ALEXANDRIA_S3_ACCESS_KEY,
            aws_secret_access_key=settings.ALEXANDRIA_S3_SECRET_KEY,
        )


@pytest.mark.parametrize("use_s3", [True, False])
@pytest.mark.parametrize("same_user", [True, False])
def test_dav(db, manabi, settings, s3, file_factory, use_s3, same_user):
    if use_s3:
        settings.ALEXANDRIA_FILE_STORAGE = "alexandria.storages.backends.s3.S3Storage"
    user = "admin"
    group = "admin"
    if same_user:
        user = "foobar"
        group = "foobar"
    with mock_aws():
        s3.create_bucket(Bucket=settings.ALEXANDRIA_S3_BUCKET_NAME)

        content_file = ContentFile(b"hello world", name="test.txt")
        file = file_factory(
            name="test.txt", content=content_file, size=content_file.size
        )
        file.modified_by_user = user
        file.modified_by_group = group
        file.save()
        dav_app = TestApp(get_dav())
        resp = dav_app.get(file.get_webdav_url("foobar", "foobar"))
        assert resp.status_code == status.HTTP_200_OK
        assert resp.body == b"hello world"

        resp = dav_app.put(file.get_webdav_url("foobar", "foobar"), b"foo bar")
        assert resp.status_code == status.HTTP_204_NO_CONTENT

        file.refresh_from_db()
        new_file = file
        document_files_count = 2
        if not same_user:
            document_files_count = 4
            new_file = (
                File.objects.filter(
                    document=file.document, variant=File.Variant.ORIGINAL
                )
                .exclude(pk=file.pk)
                .first()
            )
            assert file.modified_by_user == file.modified_by_group == "admin"

        assert file.document.files.count() == document_files_count
        assert new_file.size == new_file.content.size == 7
        assert new_file.modified_by_user == "foobar"
        new_file.content.seek(0)
        assert new_file.content.read() == b"foo bar"


@pytest.mark.freeze_time("1970-01-01")
def test_dav_propfind(db, manabi, file_factory, snapshot):
    content_file = ContentFile(b"hello world", name="test.txt")
    file = file_factory(name="test.txt", content=content_file, size=content_file.size)
    dav_app = TestApp(get_dav())
    url = file.get_webdav_url("foobar", "foobar")
    req = TestRequest.blank(url, method="PROPFIND", headers={"Depth": "1"})
    resp = dav_app.do_request(req)
    assert minidom.parseString(resp.body).toprettyxml(indent="  ") == snapshot


def test_dav_not_found(db, settings):
    key = Key.from_dictionary({"manabi": {"key": settings.MANABI_SHARED_KEY}})
    payload = ("username", "groupname", str(uuid4()))
    token = Token(key, Path("/"), payload=payload)

    assert (
        AlexandriaProvider("/").get_file_resource(
            "/",
            environ={"manabi.token": token, "wsgidav.provider": AlexandriaProvider},
            _=None,
        )
        is None
    )


def test_dav_file_infection(db, manabi, mocker, file_factory):
    mocker.patch(
        "alexandria.dav_provider.validate_file_infection",
        return_value=None,
        side_effect=ValidationError("File is infected with malware.", code="infected"),
    )

    content_file = ContentFile(b"hello world", name="test.txt")
    file = file_factory(name="test.txt", content=content_file, size=content_file.size)
    dav_app = TestApp(get_dav())
    url = file.get_webdav_url("foobar", "foobar")
    dav_app.put(url, b"foo bar", status=HTTP_FORBIDDEN)


@pytest.mark.parametrize("use_manabi", [True, False])
def test_dav_view(manabi, admin_client, document, file_factory, use_manabi, settings):
    settings.ALEXANDRIA_USE_MANABI = use_manabi
    content_file = ContentFile(b"hello world", name="test.txt")
    file = file_factory(name="test.txt", content=content_file, size=content_file.size)
    document.files.add(file)

    url = reverse("webdav-detail", args=[document.pk])
    response = admin_client.get(url)
    assert response.status_code == HTTP_200_OK
    dav_url = response.json()["data"]["attributes"]["webdav-url"]
    if use_manabi:
        assert dav_url.startswith("http://testserver/dav/")
        assert dav_url.endswith("test.txt")
    else:
        assert dav_url is None
