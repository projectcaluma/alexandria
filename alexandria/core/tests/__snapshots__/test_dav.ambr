# serializer version: 1
# name: test_dav_propfind
  '''
  <?xml version="1.0" ?>
  <ns0:multistatus xmlns:ns0="DAV:">
    <ns0:response>
      <ns0:href>/dav/test.txt</ns0:href>
      <ns0:propstat>
        <ns0:prop>
          <ns0:resourcetype/>
          <ns0:creationdate>1970-01-01T00:00:00Z</ns0:creationdate>
          <ns0:getcontentlength>11</ns0:getcontentlength>
          <ns0:getcontenttype>text/plain</ns0:getcontenttype>
          <ns0:getlastmodified>Thu, 01 Jan 1970 00:00:00 GMT</ns0:getlastmodified>
          <ns0:displayname>test.txt</ns0:displayname>
          <ns0:lockdiscovery/>
          <ns0:supportedlock>
            <ns0:lockentry>
              <ns0:lockscope>
                <ns0:exclusive/>
              </ns0:lockscope>
              <ns0:locktype>
                <ns0:write/>
              </ns0:locktype>
            </ns0:lockentry>
            <ns0:lockentry>
              <ns0:lockscope>
                <ns0:shared/>
              </ns0:lockscope>
              <ns0:locktype>
                <ns0:write/>
              </ns0:locktype>
            </ns0:lockentry>
          </ns0:supportedlock>
        </ns0:prop>
        <ns0:status>HTTP/1.1 200 OK</ns0:status>
      </ns0:propstat>
    </ns0:response>
  </ns0:multistatus>
  
  '''
# ---
